trantor/utils/Utilities.cc

```diff
@@ -428,7 +428,10 @@ static bool systemRandomBytes(void *ptr, size_t size)
       (__GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 25))))
     return getentropy(ptr, size) != -1;
 #elif defined(_WIN32)  // Windows
-    return RtlGenRandom(ptr, (ULONG)size);
+    static HCRYPTPROV handle = 0;
+    if (!handle && !CryptAcquireContext(&handle, 0, 0, PROV_RSA_FULL, CRYPT_VERIFYCONTEXT | CRYPT_SILENT))
+        return false;
+    return CryptGenRandom(handle, size, (BYTE*)ptr);
```

Comment:
Important:  This API is deprecated. New and existing software should start using [Cryptography Next Generation APIs.](https://learn.microsoft.com/en-us/windows/desktop/SecCNG/cng-portal) Microsoft may remove this API in future releases.

The above are the tips about this function on Microsoft's documentation website. Is there a better solution?

Reply:
Actually, maybe I don't need this patch? Somehow, my linking error magically disappeared...

Reply:
Also this can be raced. `static HCRYPTPROV handle` either has to be `thread_local` or we need a mutex. It is possible for 2 threads to enter this block of code at the exact same time. The C++ standard guarantees static variable initialization is MT-safe. But the following check is not. It is possible for both threads to find `handle == 0` then both calls `CryptAcquireContext`. Then both writes to `handle`.